### ISSUE TITLE
Laminar serving ELF binary as gzip data

### ISSUE THREAD

**Issue by Justasic** [ohwgiles/laminar#173]
Hello, I am a rather novice user and tried to look through previous open/closed issues for a solution to this weird issue but found nothing related. Both chrome and firefox reported that the server was returning the "wrong encoding", httpie says it failed to decode the gzipped data so I decided to make a manual request using telnet to see what it was sending and received this:
![photo_2022-08-26_03-45-32](https://user-images.githubusercontent.com/579022/186887551-ee82f23d-ec4a-46b7-98a7-8c80c4205d76.jpg)

Running `head /usr/bin/laminard` returns identical output (I know most of it is the ELF header) so I assume it's somehow sending itself(?) but I am not entirely sure how I should debug this. Laminard doesn't crash and I can run jobs just fine as well as use curl to watch the output of the jobs in the console, it's only when viewing the web interface that this occurs.

Arch Linux
Laminar AUR package version 1.2-1
Cap'n'proto version: 0.10.1-1

**Comment by ohwgiles** [comment]
I cannot reproduce this.
- Are you absolutely sure it is laminar itself that is serving this?
- I guess you changed `LAMINAR_BIND_HTTP` such that it is serving on port `6875`?
- Do you get the same result at other static laminar endpoints, such as `/style.css`, `/js/app.js`?
- Can you post the build output of the AUR package? It would be odd, but maybe could be caused by a failure in the embedding of static assets at configure time.
- Is the `gzip` binary present where you built the AUR package?

**Comment by Justasic** [comment]
> Are you absolutely sure it is laminar itself that is serving this?

Yes, I changed `LAMINAR_BIND_HTTP` and verified the `laminard` process bound to that port with `netstat -tupln`.

> I guess you changed LAMINAR_BIND_HTTP such that it is serving on port 6875?

Correct, though I did rebind the port to other port numbers to verify that was not the issue. (trying both 8080 and 8000)

> Do you get the same result at other static laminar endpoints, such as /style.css, /js/app.js?

After testing, yes. It gives the exact same error

> Can you post the build output of the AUR package? It would be odd, but maybe could be caused by a failure in the embedding of static assets at configure time.

[Sure.](https://github.com/ohwgiles/laminar/files/9468710/build_output.txt)

> Is the gzip binary present where you built the AUR package?

Yes, It's a standard arch user environment, I did verify the gzip executable exists and functions properly.


Additionally, Since I know the pain of "works for me", I used a tool called [libtree](https://github.com/haampie/libtree) to resolve all dynamic libraries used and place them in the `executables` directory in the attached zip, hopefully this can help replicate the issue. 
The zip also includes the build products of makepkg. I have tried to rebuild laminar with both clang and gcc but got the same result. I will see if I can find another arch linux system to attempt this build on, hopefully find what is wrong here.
I did test external browsers with this version of laminar (thinking maybe the browser was messed up) but those browsers returned the same error.


[build_output.txt](https://github.com/ohwgiles/laminar/files/9468710/build_output.txt)
[laminar.zip](https://github.com/ohwgiles/laminar/files/9468745/laminar.zip)

**Comment by ohwgiles** [comment]
Thanks for the detail.

Can you please try rebuilding your AUR package without stripping debug symbols? Adding `OPTIONS=(!strip)` to the `PKGBUILD` should do this, and you can check whether the binary is stripped using `nm`.

Does the unstripped version also exhibit this behaviour? If so, please send it to me too.

**Comment by Justasic** [comment]
Apologies for the delay, life's been hectic.

Your mention of the debug symbols reminded me that I have some pretty aggressive optimizations I did to my makepkg.conf, one of which was using the [mold linker](https://github.com/rui314/mold) instead of regular GNU ld. It seems that attempting to link laminar with mold causes this behavior. I tested with all the flags below but removed mold and laminar rendered the page as expected.

Since mold is still relatively new and this feels out of scope of the project overall, I'll go ahead and close the issue. The flags I was using if you would like to replicate it were the following:

```
CFLAGS="-march=native -mtune=native -O3 -pipe -fno-plt -ftree-vectorize -D_FORTIFY_SOURCE=2"
CXXFLAGS="-march=native -mtune=native -O3 -pipe -fno-plt -ftree-vectorize -D_FORTIFY_SOURCE=2"
LDFLAGS="-Wl,-O3,--sort-common,--as-needed,-z,relro,-z,now,--strip-debug,-flto -fuse-ld=mold"
OPTIONS=(strip [...])
STRIP_BINARIES="--strip-all"
STRIP_SHARED="--strip-unneeded"
STRIP_STATIC="--strip-debug"
```

Thanks for your time and helping me debug the issue!
